components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,524 @@
+        try {
+            Key privateKey = privateKeys.get(tenantId);
+            if (privateKey == null) {
+                KeyStoreManager tenantKSM = getKeyStoreManager(tenantId);
+
+                if (!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+                    /* Derive key store name */
+                    String keyStoreName = tenantDomain.trim().replace(".", "-");
+                    String jksName = keyStoreName + ".jks";
+                    /* Obtain private key */
+                    privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+                } else {
+                    try {
+                        privateKey = tenantKSM.getDefaultPrivateKey();
+                    } catch (Exception e) {
+                        String error = "Error while obtaining private key for super tenant";
+                        log.error(error, e);
+                        throw new AppManagementException(error);
+                    }
+                }
+                if (privateKey != null) {
+                    privateKeys.put(tenantId, privateKey);
+                }
+            }
+            return privateKey;
+        } catch (AppManagementException e) {
+            String error = "Error in obtaining tenant's" + tenantId + "private key";
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        }
+    }
+
+    private KeyStoreManager getKeyStoreManager(int tenantId) throws AppManagementException {
+        try {
+            AppManagerUtil.loadTenantRegistry(tenantId);
+            return KeyStoreManager.getInstance(tenantId);
+        } catch (AppManagementException e) {
+            String error = "Error in obtaining  key store manager for tenant " + tenantId;
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        }
+    }
+
+    /**
+     * Helper method to add public certificate to JWT_HEADER to signature verification.
+     *
+     * @param endUserName
+     * @return jwt header as a string
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private String addThumbPrintToHeader(String endUserName) throws AppManagementException {
+        int tenantId = getTenantId(endUserName);
+        try {
+            StringBuilder jwtHeader = new StringBuilder();
+            String base64EncodedThumbPrint = getBase64EncodedThumbPrint(endUserName, tenantId);
+            if (base64EncodedThumbPrint == null) {
+                log.error("Base64 encoded thumb print is null for tenant : " + tenantId);
+            }
+            jwtHeader.append("\"x5t\":\"");
+            jwtHeader.append(base64EncodedThumbPrint);
+            jwtHeader.append("\"");
+            return jwtHeader.toString();
+        } catch (Exception e) {
+            String error = "Error in adding tenant's" + tenantId + "public certificate";
+            throw new AppManagementException(error);
```

Comment:
In above methods, you have logged and throws the exception but here. you have only throw the exception and also error trace get dropped on the way. Please fix it.